core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java

```diff
@@ -184,6 +185,81 @@ static <T> StreamMessage<T> of(Publisher<? extends T> publisher) {
+     *
+     * @param stage the {@link CompletionStage} which will produce the actual {@link StreamMessage}
+     */
+    static <T> StreamMessage<T> of(CompletionStage<? extends Publisher<? extends T>> stage) {
+        requireNonNull(stage, "stage");
+
+        if (stage instanceof CompletableFuture) {
+            return of((CompletableFuture<? extends Publisher<? extends T>>) stage);
+        } else {
+            final DeferredStreamMessage<T> deferred = new DeferredStreamMessage<>();
+            //noinspection unchecked
+            deferred.delegateWhenCompleteStage((CompletionStage<? extends Publisher<T>>) stage);
+            return deferred;
+        }
+    }
+
+    /**
+     * Creates a new {@link StreamMessage} that delegates to the {@link StreamMessage} produced by the specified
+     * {@link CompletionStage}. If the specified {@link CompletionStage} fails, the returned
+     * {@link StreamMessage} will be closed with the same cause as well.
+     *
+     * @param future the {@link CompletionStage} which will produce the actual {@link StreamMessage}
+     */
+    static <T> StreamMessage<T> of(CompletableFuture<? extends Publisher<? extends T>> future) {
+        requireNonNull(future, "future");
+
+        if (future.isDone()) {
+            if (!future.isCompletedExceptionally()) {
+                final Publisher<? extends T> publisher = future.getNow(null);
+                if (publisher instanceof StreamMessage) {
+                    //noinspection unchecked
+                    return (StreamMessage<T>) publisher;
+                }
+                return new PublisherBasedStreamMessage<>(publisher);
+            }
+
+            try {
+                future.join();
+                // Should never reach here.
+                throw new Error();
+            } catch (Throwable cause) {
+                return aborted(Exceptions.peel(cause));
+            }
+        }
+
+        final DeferredStreamMessage<T> deferred = new DeferredStreamMessage<>();
+        //noinspection unchecked
+        deferred.delegateWhenCompleteStage((CompletionStage<? extends Publisher<T>>) future);
+        return deferred;
+    }
+
+    /**
+     * Creates a new {@link StreamMessage} that delegates to the {@link StreamMessage} produced by the specified
+     * {@link CompletionStage}. If the specified {@link CompletionStage} fails, the returned
+     * {@link StreamMessage} will be closed with the same cause as well.
+     *
+     * @param stage the {@link CompletionStage} which will produce the actual {@link StreamMessage}
+     * @param subscriberExecutor the {@link EventExecutor} which will be used when a user subscribes
+     *                           the returned {@link StreamMessage} using {@link #subscribe(Subscriber)}
+     *                           or {@link #subscribe(Subscriber, SubscriptionOption...)}.
+     */
+    static <T> StreamMessage<T> of(CompletionStage<? extends StreamMessage<? extends T>> stage,
```

Comment:
I think optionally we could also apply the CF completion check here too (I find it odd that only some APIs do this optimization, while others don't)

Reply:
If so, we cannot use the executor specified. 😉 Let me add a comment there.